session/session.go

```diff
@@ -0,0 +1,113 @@
+package session
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"launchpad.net/gozk"
+)
+
+var (
+	ErrZkSessionNotConnected       = fmt.Errorf("Unable to connect to ZooKeeper.")
+	ErrZkSessionTimeout            = fmt.Errorf("Session connection timeout expired.")
+	ErrZkSessionExpired            = fmt.Errorf("Session expired.")
+	ErrZkSessionEventChannelClosed = fmt.Errorf("Session event channel closed.")
+)
+
+type ZkSession struct {
+	servers     string
+	dialTimeout time.Duration
+	recvTimeout time.Duration
+	Connection  *zookeeper.Conn
+	events      <-chan zookeeper.Event
+	mu          sync.Mutex
+}
+
+func NewZkSession(servers string, dialTimeout time.Duration, recvTimeout time.Duration) (*ZkSession, error) {
+	zkconn, zkevents, err := zookeeper.Dial(servers, recvTimeout)
+
+	if err != nil {
+		return nil, err
+	}
+
+	s := &ZkSession{servers: servers, dialTimeout: dialTimeout, recvTimeout: recvTimeout, mu: sync.Mutex{}}
+
+	if err := s.establish(zkconn, zkevents); err != nil {
+		return nil, err
+	}
+
+	go s.manage()
+
+	return s, nil
+}
+
+func (s *ZkSession) establish(zkconn *zookeeper.Conn, zkevents <-chan zookeeper.Event) error {
+	defer s.mu.Unlock()
+	s.mu.Lock()
+
+	for {
+		select {
+		case event, ok := <-zkevents:
+			if !ok {
+				return ErrZkSessionEventChannelClosed
+			}
+
+			if event.State == zookeeper.STATE_EXPIRED_SESSION {
+				_ = zkconn.Close()
+				return ErrZkSessionExpired
+			}
+
+			// Only the STATE_CONNECTED session event is valid when establishing a new session.
```

Comment:
the one about invalid auth, for example, does kind of warrant different handling. (bail?)